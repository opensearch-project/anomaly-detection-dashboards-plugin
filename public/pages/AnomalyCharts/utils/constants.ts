/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export const INITIAL_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0.0,
  maxAnomalyGrade: 0.0,
  minConfidence: 0.0,
  maxConfidence: 0.0,
  lastAnomalyOccurrence: '',
};

export enum CHART_FIELDS {
  PLOT_TIME = 'plotTime',
  ANOMALY_GRADE = 'anomalyGrade',
  CONFIDENCE = 'confidence',
  DATA = 'data',
  AGG_INTERVAL = 'aggInterval',
  EXPECTED_VALUE = 'expectedValue',
}

export enum CHART_COLORS {
  ANOMALY_GRADE_COLOR = '#D13212',
  FEATURE_DATA_COLOR = '#16191F',
  FEATURE_COLOR = '#fcd529',
  CONFIDENCE_COLOR = '#017F75',
  LIGHT_BACKGROUND = '#FFFFFF',
  DARK_BACKGROUND = '#1D1E24',
}

export const FEATURE_CHART_THEME = [
  {
    lineSeriesStyle: {
      line: {
        strokeWidth: 2,
        visible: true,
        opacity: 0.5,
      },
      point: {
        visible: true,
        stroke: CHART_COLORS.FEATURE_DATA_COLOR,
      },
    },
  },
  {
    colors: {
      vizColors: [CHART_COLORS.FEATURE_DATA_COLOR],
    },
  },
];

export const ANOMALY_CHART_THEME = [
  {
    colors: {
      vizColors: [
        CHART_COLORS.CONFIDENCE_COLOR,
        CHART_COLORS.ANOMALY_GRADE_COLOR,
      ],
    },
  },
];

export const DATE_PICKER_QUICK_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
  REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
  MONITORING_INTERVALS = 60,
}

export const DEFAULT_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0,
  maxAnomalyGrade: 0,
  minConfidence: 0,
  maxConfidence: 0,
  lastAnomalyOccurrence: '-',
};

export const HEATMAP_CHART_Y_AXIS_WIDTH = 30;
