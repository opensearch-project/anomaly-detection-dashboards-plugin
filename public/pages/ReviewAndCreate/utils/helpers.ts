/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  FEATURE_TYPE,
  FILTER_TYPES,
  UNITS,
  FeatureAttributes,
  Detector,
  UiFeature,
} from '../../../models/interfaces';
import moment from 'moment';
import { get, isEmpty, snakeCase } from 'lodash';
import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
import { FeaturesFormikValues } from '../../ConfigureModel/models/interfaces';
import { CreateDetectorFormikValues } from '../../CreateDetectorSteps/models/interfaces';
import { OPERATORS_QUERY_MAP } from '../../DefineDetector/utils/whereFilters';
import { convertTimestampToNumber } from '../../../utils/utils';
import { CUSTOM_AD_RESULT_INDEX_PREFIX } from '../../../../server/utils/constants';
import { formikToImputationOption, formikToRules } from '../../ConfigureModel/utils/helpers';

export function formikToDetector(values: CreateDetectorFormikValues): Detector {
  const detectionDateRange = values.historical
    ? {
        startTime: convertTimestampToNumber(values.startTime),
        endTime: convertTimestampToNumber(values.endTime),
      }
    : undefined;
  var resultIndex = values.resultIndex;
  if (resultIndex && resultIndex.trim().length > 0) {
    resultIndex = CUSTOM_AD_RESULT_INDEX_PREFIX + resultIndex;
  }
  let detectorBody = {
    name: values.name,
    description: values.description,
    indices: formikToIndices(values.index),
    resultIndex: resultIndex,
    filterQuery: formikToFilterQuery(values),
    uiMetadata: {
      features: { ...featuresToUIMetadata(values.featureList) },
      filters: get(values, 'filters', []),
    },
    featureAttributes: formikToFeatureAttributes(values.featureList),
    timeField: values.timeField,
    detectionInterval: {
      period: { interval: values.interval, unit: UNITS.MINUTES },
    },
    windowDelay: {
      period: { interval: values.windowDelay, unit: UNITS.MINUTES },
    },
    shingleSize: values.shingleSize,
    categoryField: !isEmpty(values?.categoryField)
      ? values.categoryField
      : undefined,
    resultIndexMinAge:
      resultIndex && resultIndex.trim().length > 0
        ? values.resultIndexMinAge
        : undefined,
    resultIndexMinSize:
      resultIndex && resultIndex.trim().length > 0
        ? values.resultIndexMinSize
        : undefined,
    resultIndexTtl:
      resultIndex && resultIndex.trim().length > 0
        ? values.resultIndexTtl
        : undefined,
    flattenCustomResultIndex:
      resultIndex && resultIndex.trim().length > 0
        ? values.flattenCustomResultIndex
        : undefined,
    imputationOption: formikToImputationOption(values.imputationOption),
    rules: formikToRules(values.suppressionRules),
  } as Detector;

  // Optionally add detection date range
  if (detectionDateRange) {
    detectorBody = {
      ...detectorBody,
      //@ts-ignore
      detectionDateRange: detectionDateRange,
    };
  }

  return detectorBody;
}

export const formikToIndices = (indices: { label: string }[]) =>
  indices.map((index) => index.label);

export const formikToFilterQuery = (
  values: CreateDetectorFormikValues | DetectorDefinitionFormikValues
) => {
  let filterQuery = {};
  const filters = get(values, 'filters', []);

  // If we have filters: need to combine into a single filter query.
  // Need to handle each filter type differently when converting
  if (filters.length > 0) {
    let filterArr = [] as any[];
    filters.forEach((filter) => {
      if (filter.filterType === FILTER_TYPES.SIMPLE) {
        filterArr.push(
          //@ts-ignore
          OPERATORS_QUERY_MAP[filter.operator]?.query(filter)
        );
      } else {
        filterArr.push(
          //@ts-ignore
          JSON.parse(filter.query)
        );
      }
    });
    filterQuery = {
      bool: {
        filter: [...filterArr],
      },
    };
  } else {
    filterQuery = { match_all: {} };
  }
  return filterQuery;
};

export function formikToFeatureAttributes(
  values: FeaturesFormikValues[],
  forPreview: boolean = false
): FeatureAttributes[] {
  //@ts-ignore
  return values.map(function (value) {
    const id = forPreview
      ? value.featureId
      : value.newFeature
      ? undefined
      : value.featureId;
    return {
      featureId: id,
      featureName: value.featureName,
      featureEnabled: value.featureEnabled,
      importance: 1,
      aggregationQuery: formikToAggregation(value),
    };
  });
}

export function featuresToUIMetadata(values: FeaturesFormikValues[]) {
  // TODO:: Delete Stale metadata if name is changed
  let features: {
    [key: string]: UiFeature;
  } = {};
  values.forEach((value) => {
    if (value.featureType === FEATURE_TYPE.SIMPLE) {
      features[value.featureName] = {
        featureType: value.featureType,
        aggregationBy: value.aggregationBy,
        aggregationOf:
          value.aggregationOf && value.aggregationOf.length
            ? value.aggregationOf[0].label
            : undefined,
      };
    } else {
      features[value.featureName] = {
        featureType: value.featureType,
      };
    }
  });
  return features;
}

export function formikToAggregation(values: FeaturesFormikValues) {
  if (values.featureType === FEATURE_TYPE.SIMPLE) {
    return values.aggregationBy &&
      values.aggregationOf &&
      values.aggregationOf.length > 0
      ? {
          [snakeCase(values.featureName)]: {
            [values.aggregationBy]: { field: values.aggregationOf[0].label },
          },
        }
      : {};
  }
  return JSON.parse(values.aggregationQuery);
}

export function toStringConfigCell(obj: any): string {
  if (typeof obj != 'undefined') {
    if (obj.hasOwnProperty('period')) {
      let period = obj.period;
      return period.interval + ' ' + period.unit;
    } else if (typeof obj == 'number') {
      // epoch
      return moment(obj).format('MM/DD/YY hh:mm A');
    }
  }
  return '-';
}
