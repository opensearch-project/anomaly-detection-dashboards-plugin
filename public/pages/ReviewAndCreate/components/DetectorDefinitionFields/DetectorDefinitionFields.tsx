/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import {
  EuiFlexGrid,
  EuiFlexItem,
  EuiButton,
  EuiCallOut,
  EuiLoadingSpinner,
  EuiFlexGroup,
  EuiText,
} from '@elastic/eui';
import React from 'react';
import { get, isEqual } from 'lodash';
import {
  Detector,
  ValidationSettingResponse,
} from '../../../../models/interfaces';
import { FilterDisplayList } from '../FilterDisplayList';
import { ConfigCell, FixedWidthRow } from '../../../../components/ConfigCell';
import { toStringConfigCell } from '../../utils/helpers';
interface DetectorDefinitionFieldsProps {
  detector: Detector;
  onEditDetectorDefinition(): void;
  isCreate: boolean;
  validationError?: boolean;
  validDetectorSettings?: boolean;
  validationResponse?: ValidationSettingResponse;
  isLoading?: boolean;
  isCreatingDetector?: boolean;
}

export const DetectorDefinitionFields = (
  props: DetectorDefinitionFieldsProps
) => {
  const filterInputs = {
    uiMetadata: get(props, 'detector.uiMetadata', {}),
    filterQuery: JSON.stringify(
      get(props, 'detector.filterQuery', {}) || {},
      null,
      4
    ),
  };

  const getValidationCallout = () => {
    //When validation response is loading then displaying loading spinner, don't display
    // after clicking on "create detector" button as isLoading will be true from that request
    if (props.isLoading && !props.isCreatingDetector) {
      return (
        <EuiCallOut
          title={
            <div>
              <EuiFlexGroup direction="row" gutterSize="xs">
                <EuiLoadingSpinner size="l" style={{ marginRight: '12px' }} />
                <EuiText>
                  <p>Validating detector configurations</p>
                </EuiText>
              </EuiFlexGroup>
            </div>
          }
          style={{ marginBottom: '10px' }}
          size="s"
          color="primary"
        />
      );
    }
    // Callouts only displayed based on if validDetectorSettings is true or not, referring to
    // to response content from validation API (empty body or response issue body).
    // validationError refers to if there was an exception from validation API
    // such as a security exception or error establishing network connection on request.
    // This means no callout will be displayed since validation wasn't able to say if settings are valid or not.
    if (props.validationResponse != undefined && !props.validationError) {
      if (props.validDetectorSettings) {
        return (
          <EuiCallOut
            title="Detector settings are validated"
            color="success"
            iconType="check"
            size="s"
            style={{ marginBottom: '10px' }}
          ></EuiCallOut>
        );
      } else if (
        !props.validDetectorSettings &&
        props.validationResponse.hasOwnProperty('message')
      ) {
        if (
          isEqual(get(props, 'validationResponse.validationType', ''), 'model')
        ) {
          return (
            <EuiCallOut
              title="We identified some areas that might improve your model"
              color="warning"
              iconType="iInCircle"
              size="s"
              style={{ marginBottom: '10px' }}
            >
              {JSON.stringify(props.validationResponse.message).replace(
                /\"/g,
                ''
              )}
            </EuiCallOut>
          );
        } else {
          return (
            <EuiCallOut
              title="Issues found in the detector settings"
              color="danger"
              iconType="alert"
              size="s"
              style={{ marginBottom: '10px' }}
            >
              <ul>
                <li>{props.validationResponse.message}</li>
              </ul>
            </EuiCallOut>
          );
        }
      } else {
        return null;
      }
    }
  };

  const minAge = get(props, 'detector.resultIndexMinAge', '-');
  const minSize = get(props, 'detector.resultIndexMinSize', '-');
  const ttl = get(props, 'detector.resultIndexTtl', '-');

  return (
    <ContentPanel
      title="Detector settings"
      titleDataTestSubj="detectorSettingsHeader"
      titleSize="s"
      panelStyles={{ margin: '0px' }}
      actions={[
        <EuiButton
          data-test-subj="editDetectorSettingsButton"
          onClick={props.onEditDetectorDefinition}
        >
          Edit
        </EuiButton>,
      ]}
    >
      {props.isCreate ? getValidationCallout() : null}
      <EuiFlexGrid columns={0} gutterSize="l" style={{ border: 'none' }}>
        <EuiFlexItem data-test-subj="detectorNameCell">
          <ConfigCell
            title="Name"
            description={get(props, 'detector.name', '')}
          />
        </EuiFlexItem>
        <EuiFlexItem data-test-subj="indexNameCell">
          <ConfigCell
            title="Data source index"
            description={get(props, 'detector.indices.0', '')}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <FixedWidthRow label="Data filter">
            <FilterDisplayList {...filterInputs} />
          </FixedWidthRow>
        </EuiFlexItem>
        <EuiFlexItem>
          <ConfigCell
            title="Detector interval"
            description={toStringConfigCell(
              get(props, 'detector.detectionInterval', 0)
            )}
          />
        </EuiFlexItem>
        {props.isCreate ? null : (
          <EuiFlexItem data-test-subj="detectorIdCell">
            <ConfigCell
              title="ID"
              description={get(props, 'detector.id', '')}
            />
          </EuiFlexItem>
        )}
        <EuiFlexItem data-test-subj="detectorDescriptionCell">
          <ConfigCell
            title="Description"
            description={get(props, 'detector.description', '')}
          />
        </EuiFlexItem>
        <EuiFlexItem data-test-subj="timestampNameCell">
          <ConfigCell
            title="Timestamp"
            description={get(props, 'detector.timeField', '')}
          />
        </EuiFlexItem>
        {props.isCreate ? null : (
          <EuiFlexItem>
            <ConfigCell
              title="Last Updated"
              description={toStringConfigCell(
                get(props, 'detector.lastUpdateTime', '')
              )}
            />
          </EuiFlexItem>
        )}
        <EuiFlexItem>
          <ConfigCell
            title="Window delay"
            description={toStringConfigCell(
              get(props, 'detector.windowDelay', 0)
            )}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <ConfigCell
            title="Custom result index"
            description={get(props, 'detector.resultIndex', '-')}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <ConfigCell
            title="Custom result index min age"
            description={minAge === '-' ? minAge : `${minAge} Days`}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <ConfigCell
            title="Custom result index min size"
            description={minSize == '-' ? minSize : `${minSize} MB`}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <ConfigCell
            title="Custom result index TTL"
            description={ttl == '-' ? ttl : `${ttl} Days`}
          />
        </EuiFlexItem>
      </EuiFlexGrid>
    </ContentPanel>
  );
};
