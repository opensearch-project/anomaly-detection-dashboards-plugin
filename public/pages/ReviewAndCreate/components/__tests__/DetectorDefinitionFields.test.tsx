/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { DetectorDefinitionFields } from '../DetectorDefinitionFields/DetectorDefinitionFields';
import {
  Detector,
  UNITS,
  FILTER_TYPES,
  OPERATORS_MAP,
} from '../../../../models/interfaces';
import { Formik } from 'formik';
import { DATA_TYPES } from '../../../../utils/constants';

const testDetector = {
  id: 'test-id',
  name: 'test-detector',
  indices: ['test-index'],
  detectionInterval: {
    period: {
      interval: 10,
      unit: UNITS.MINUTES,
    },
  },
  description: 'test-description',
  timeField: 'test-timefield',
  windowDelay: {
    period: {
      interval: 1,
      unit: UNITS.MINUTES,
    },
  },
  uiMetadata: {
    filters: [
      {
        filterType: FILTER_TYPES.SIMPLE,
        fieldInfo: [
          {
            label: 'test-filter-field',
            type: DATA_TYPES.TEXT,
          },
        ],
        operator: OPERATORS_MAP.IS,
        fieldValue: 'null',
      },
    ],
  },
} as Detector;

describe('<AdditionalSettings /> spec', () => {
  test('renders the component in create mode (no ID)', () => {
    const onEditDetectorDefinition = jest.fn();
    const { container, getByText, queryByText } = render(
      <Formik initialValues={{}} onSubmit={jest.fn()}>
        {() => (
          <div>
            <DetectorDefinitionFields
              onEditDetectorDefinition={onEditDetectorDefinition}
              detector={testDetector}
              isCreate={true}
            />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
    getByText('test-detector');
    getByText('test-index');
    getByText('test-filter-field is null');
    getByText('10 Minutes');
    getByText('test-description');
    getByText('test-timefield');
    getByText('1 Minutes');
    expect(queryByText('test-id')).toBeNull();
  });
  test('renders the component in edit mode (with ID)', () => {
    const onEditDetectorDefinition = jest.fn();
    const { container, getByText, queryByText } = render(
      <Formik initialValues={{}} onSubmit={jest.fn()}>
        {() => (
          <div>
            <DetectorDefinitionFields
              onEditDetectorDefinition={onEditDetectorDefinition}
              detector={testDetector}
              isCreate={false}
            />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
    getByText('test-detector');
    getByText('test-index');
    getByText('test-filter-field is null');
    getByText('10 Minutes');
    getByText('test-description');
    getByText('test-timefield');
    getByText('1 Minutes');
    expect(queryByText('test-id')).not.toBeNull();
  });
});
