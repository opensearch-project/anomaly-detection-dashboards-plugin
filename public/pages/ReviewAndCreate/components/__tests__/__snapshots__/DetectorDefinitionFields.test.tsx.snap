// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdditionalSettings /> spec renders the component in create mode (no ID) 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px; margin: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Detector settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <button
              class="euiButton euiButton--primary"
              data-test-subj="editDetectorButton"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Edit
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--wrap euiFlexGrid--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Name
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-detector
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Data source index
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-index
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Data filter
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <ol>
                  <li
                    class="enabled"
                  >
                    test-filter-field is null
                  </li>
                </ol>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Detector interval
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    10 Minutes
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Description
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-description
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Timestamp
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-timefield
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Window delay
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    1 Minutes
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AdditionalSettings /> spec renders the component in edit mode (with ID) 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px; margin: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Detector settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem content-panel-subTitle"
            style="line-height: normal; max-width: 75%;"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <button
              class="euiButton euiButton--primary"
              data-test-subj="editDetectorButton"
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Edit
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--s"
    />
    <div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
      />
      <div
        style="padding: 10px 0px;"
      >
        <div
          class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--wrap euiFlexGrid--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Name
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-detector
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Data source index
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-index
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Data filter
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <ol>
                  <li
                    class="enabled"
                  >
                    test-filter-field is null
                  </li>
                </ol>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Detector interval
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    10 Minutes
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  ID
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-id
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Description
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-description
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Timestamp
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    test-timefield
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Last Updated
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    -
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="random_html_id-row"
              style="width: 250px;"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  Window delay
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiText euiText--medium"
                  id="random_html_id"
                >
                  <p
                    class="enabled"
                  >
                    1 Minutes
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
